### PR TITLE
Optimize the merge if all hashSets are singleLevel in UniqExactSet

### PR DESCRIPTION
**Pull request by jiebinn**
In PR(https://github.com/ClickHouse/ClickHouse/pull/50748), it has added new phase `parallelizeMergePrepare` before merge if all the hashSets are not all singleLevel or not all twoLevel. Then it will convert all the singleLevelSet to twoLevelSet in parallel, which will increase the CPU utilization and QPS.

But if all the hashtables are singleLevel, it could also benefit from the `parallelizeMergePrepare` optimization in most cases if the hashtable size are not too small. 

Then we should tune the threshold of the hashtable size. The total dataSet should not be that large. The unique column count should be proportional to the number of hits_v1. Keep the unique column count very small or we will never find the cross point in limited threads. And we would choose the dataSet `hits_v1` and the column `SearchPhase`. We have to make sure all the hashtable are singleLevel no matter what is the threads_num. Also, we have to make sure there is a cross point in the figure.

By tuning the Query `SELECT COUNT(DISTINCT SearchPhase) FROM hits_v1` in different threads, we have got the mild threshold 6,000 (total_hashtable_size/hashtable_num). `COUNT(DISTINCT SearchPhase)` is only 132,256. Even when threads_num is very small, the hashtable size could not reach the threshold 100,000 and they are all singleLevel.
image

Test patch with the Query 'SELECT COUNT(DISTINCT Title) FROM hits_v1' on 2x80 vCPUs server. If the threads are less than 48, the hashSets are all twoLevel or mixed by singleLevel and twoLevel. If the threads are over 56, all the hashSets are singleLevel. And the QPS has got at most **2.35x** performance gain.
image

    Threads Opt/Base
    8       100.0%
    16      99.4%
    24      110.3%
    32      99.9%
    40      99.3%
    48      99.8%
    56      183.0%
    64      234.7%
    72      233.1%
    80      229.9%
    88      224.5%
    96      229.6%
    104     235.1%
    112     229.5%
    120     229.1%
    128     217.8%
    136     222.9%
    144     217.8%
    152     204.3%
    160     203.2%

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimize the merge if all hashSets are singleLevel in UniqExactSet.

### CODE REVIEW

**Comment by jiebinn** [comment]
Hi @nickitat, the previous PR 'https://github.com/ClickHouse/ClickHouse/pull/50748' is only used when the hashSets are mixed by singleLevel and twoLevel. But I have found that if all the hashSets are singleLevel, it could also benefit from the previous patch a lot in most cases. I have submitted the additional PR in that case if all singleLevelHash.

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 1f6a6c83ccca3bc8e86a1f9128974272282ed0ce with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/52973/1f6a6c83ccca3bc8e86a1f9128974272282ed0ce/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/AggregateFunctions/UniqExactSet.h [hunk 1]
```diff
@@ -34,14 +34,21 @@ class UniqExactSet
...
         }
 
-        if (single_level_set_num > 0 && single_level_set_num < data_vec.size())
+        if (single_level_set_num == data_vec.size())
+        {
+            for (auto ele : data_vec)
+                all_single_hash_size += ele->size();
+        }
+
+        if ((single_level_set_num > 0 && single_level_set_num < data_vec.size()) || ((all_single_hash_size/data_vec.size()) > 6000))
```
Comment (nickitat) [inline src/AggregateFunctions/UniqExactSet.h:51]: pls leave here a link to your explanation

**Review by nickitat — APPROVED** [review]

**Comment by jiebinn** [comment]
The CI error `fail to generate performance report` seems not related to this patch.

**Comment by jiebinn** [comment]
Hi @nickitat , thanks for your help to review the patch. BTW, do you think what else should I do for this PR?

**Pull request merged**